contracts/periphery/src/execute.rs

```diff
@@ -1,125 +1,305 @@
-use cosmwasm_std::{attr, coin, to_binary, BankMsg, Env, MessageInfo, SubMsg, Uint128, WasmMsg};
+use cosmwasm_std::{
+    attr, coin, to_json_binary, BankMsg, Coin, CosmosMsg, Env, MessageInfo, Uint128, WasmMsg,
+};
 use cosmwasm_std::{DepsMut, Response};
-use ibcx_interface::periphery::{ExecuteMsg, SwapInfo};
+use ibcx_interface::periphery::{extract_pool_ids, ExecuteMsg, SwapInfo};
 use ibcx_interface::{core, helpers::IbcCore};
+use ibcx_pool::{query_pools, Simulator};
 
-use crate::state::{Context, CONTEXT};
-use crate::REPLY_ID_BURN_EXACT_AMOUNT_IN;
-use crate::{error::ContractError, msgs::make_mint_swap_exact_out_msgs};
+use crate::error::ContractError;
+use crate::{coin_sorter, deduct_fee, expand_fee, make_unit_converter};
 
-pub fn mint_exact_amount_out(
+pub fn mint_exact_amount_in(
     deps: DepsMut,
     env: Env,
     info: MessageInfo,
     core_addr: String,
-    output_amount: Uint128,
-    input_asset: String,
+    desired_denom: String,
+    min_index_amount: Uint128,
     swap_info: Vec<SwapInfo>,
 ) -> Result<Response, ContractError> {
     // query to core contract
     let core = IbcCore(deps.api.addr_validate(&core_addr)?);
+    let core_fee = core.get_fee(&deps.querier, None)?;
     let core_config = core.get_config(&deps.querier, None)?;
+    let core_portfolio = core.get_portfolio(&deps.querier, None)?;
 
-    // input & output
-    let max_input_amount = cw_utils::must_pay(&info, &input_asset)?;
-    let max_input = coin(max_input_amount.u128(), &input_asset);
-    let output = coin(output_amount.u128(), core_config.index_denom);
+    let desired_asset =
+        cw_utils::must_pay(&info, &desired_denom).map(|v| coin(v.u128(), &desired_denom))?;
+
+    let pool_ids = extract_pool_ids(swap_info.clone());
+    let pools = query_pools(&deps.as_ref(), pool_ids)?;
 
-    let sim_resp = core.simulate_mint(&deps.querier, output.amount, None, None)?;
-    let mut sim_amount_desired = sim_resp.fund_spent;
-    sim_amount_desired.sort_by(|a, b| a.denom.cmp(&b.denom));
+    let deps_ref = deps.as_ref();
+    let sim = Simulator::new(&deps_ref, &pools, &swap_info, &core_portfolio.units);
+    let sim_res = sim
+        .estimate_index_for_input(
+            desired_asset.clone(),
+            Some(min_index_amount),
+            Some(min_index_amount),
+            None,
+        )?
```

Comment:
ask; Is there any reason you make 2 parameters `min_index_amount` and `init_index_amount` in `estimate_index_for_input` function? Because I couldn't find any use cases in this repo.

Reply:
It's because we cannot estimate the actual swap amount in one simulation

Reply:
So we need to setup initial amount to simulate it and make closer by binary search